chain-extensions/unified-accounts/src/lib.rs

```diff
@@ -0,0 +1,106 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+// Astar is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Astar is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Astar. If not, see <http://www.gnu.org/licenses/>.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use astar_primitives::{
+    ethereum_checked::AccountMapping,
+    evm::{EvmAddress, UnifiedAddressMapper},
+};
+use core::marker::PhantomData;
+use sp_runtime::DispatchError;
+
+use frame_support::{traits::Get, DefaultNoBound};
+use pallet_contracts::chain_extension::{
+    ChainExtension, Environment, Ext, InitState, Result as DispatchResult, RetVal,
+};
+use pallet_evm::AddressMapping;
+use parity_scale_codec::Encode;
+pub use unified_accounts_chain_extension_types::Command::{self, *};
+
+#[derive(DefaultNoBound)]
+pub struct UnifiedAccountsExtension<T, UA>(PhantomData<(T, UA)>);
+
+impl<T, UA> ChainExtension<T> for UnifiedAccountsExtension<T, UA>
+where
+    T: pallet_contracts::Config + pallet_unified_accounts::Config,
+    UA: UnifiedAddressMapper<T::AccountId>,
+{
+    fn call<E>(&mut self, env: Environment<E, InitState>) -> DispatchResult<RetVal>
+    where
+        E: Ext<T = T>,
+    {
+        let mut env = env.buf_in_buf_out();
+        match env.func_id().try_into().map_err(|_| {
+            DispatchError::Other("Unsupported func id in Unified Accounts Chain Extension")
+        })? {
+            GetEvmAddress => {
+                let account_id: T::AccountId = env.read_as()?;
+
+                let base_weight = <T as frame_system::Config>::DbWeight::get().reads(1);
+                env.charge_weight(base_weight)?;
+                // write to buffer
+                UA::to_h160(&account_id).using_encoded(|r| env.write(r, false, None))?;
+            }
+            GetEvmAddressOrDefault => {
+                let account_id: T::AccountId = env.read_as()?;
+
+                let base_weight = <T as frame_system::Config>::DbWeight::get().reads(1);
+                env.charge_weight(base_weight)?;
+
+                // read the storage item
+                let mapped = UA::to_h160(&account_id);
+
+                let is_mapped = mapped.is_some();
+                let evm_address = mapped.unwrap_or_else(|| {
+                    // fallback to default account_id
+                    T::DefaultNativeToEvm::into_h160(account_id)
+                });
+                // write to buffer
+                (evm_address, is_mapped).using_encoded(|r| env.write(r, false, None))?;
```

Comment:
This doesn't look really Rust-like to me.
Returning an `enum` would be more appropriate: (`Mapped(address), Default(address)`).
That also makes it more extensible for the future (not sure how we'd need it, but in general).

Anyhow, if others are ok with the proposed solution, you can ignore this comment.

Reply:
I like the rust way too.
I'll update the PR tomorrow with changes.

cc: @PierreOssun @shaunxw

Reply:
Agree, the `enum` one looks better.

@PierreOssun the CE interface in ink! can get the decoded `enum` type without extra work in SDK right? Pls correct me if I'm wrong.

Reply:
I've updated the CE with enum, please have a look

> the CE interface in ink! can get the decoded enum type without extra work in SDK right?

Yes, as long as it supports scale decode it'll be fine, I've updated the [test ink! contracts](https://github.com/AstarNetwork/ink-test-contracts/pull/3) here.